apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -177,89 +177,107 @@ resolve_loop(Begin, Workers, Fun, FailMsg, Zone) ->
 -spec(start_worker(protocol(), upstream(), binary()) -> pid()).
 start_worker(Protocol, Upstream, Request) ->
     Pid = self(),
+    Begin = erlang:monotonic_time(),
     proc_lib:spawn(fun () ->
-        init_worker(Protocol, Pid, Upstream, Request)
+        Address = upstream_to_binary(Upstream),
+        init_worker(Protocol, Pid, {Upstream, Address}, Request),
+        prometheus_histogram:observe(
+            dns, worker_requests_duration_seconds,
+            [Address, udp], erlang:monotonic_time() - Begin)
     end).
 
--spec(init_worker(protocol(), pid(), upstream(), binary()) -> ok).
-init_worker(udp, Pid, Upstreams, Request) ->
-    udp_worker(Pid, Upstreams, Request);
-init_worker(tcp, Pid, Upstreams, Request) ->
-    tcp_worker(Pid, Upstreams, Request).
+-spec(init_worker(protocol(), pid(), {upstream(), binary()}, binary()) -> ok).
+init_worker(udp, Pid, Upstream, Request) ->
+    udp_worker(Pid, Upstream, Request);
+init_worker(tcp, Pid, Upstream, Request) ->
+    tcp_worker(Pid, Upstream, Request).
 
--spec(udp_worker(pid(), upstream(), binary()) -> ok).
-udp_worker(Pid, Upstream = {IP, Port}, Request) ->
+-spec(udp_worker(pid(), {upstream(), binary()}, binary()) -> ok).
+udp_worker(Pid, {Upstream = {IP, Port}, Address}, Request) ->
     StartTime = erlang:monotonic_time(millisecond),
     Opts = [{reuseaddr, true}, {active, once}, binary],
     case gen_udp:open(0, Opts) of
         {ok, Socket} ->
             case gen_udp:send(Socket, IP, Port, Request) of
                 ok ->
-                    udp_worker(StartTime, Pid, Socket, Upstream);
-                {error, Error} ->
-                    lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+                    udp_worker(StartTime, Pid, Socket, {Upstream, Address});
+                {error, Reason} ->
+                    lager:warning(
+                        "DNS worker ~p failed with ~p",
+                        [Address, Reason]),
+                    emit_worker_failure(Address, udp, Reason)
             end;
-        {error, Error} ->
-            lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+        {error, Reason} ->
+            lager:warning("DNS worker ~p failed with ~p", [Address, Reason]),
+            emit_worker_failure(Address, udp, Reason)
     end.
 
--spec(udp_worker(integer(), pid(), gen_udp:socket(), upstream()) -> ok).
-udp_worker(StartTime, Pid, Socket, Upstream = {IP, Port}) ->
+-spec(udp_worker(integer(), pid(), gen_udp:socket(), {upstream(), binary()}) -> ok).
+udp_worker(StartTime, Pid, Socket, {Upstream = {IP, Port}, Address}) ->
     MonRef = erlang:monitor(process, Pid),
```

Comment:
Also maybe something is still confusing for me. This Pid here is `self()` before we call spawn for the worker. So I am assuming the worker monitors its parent.

- Why is it not necessary to demonitor before the worker goes away?
- Should we really emit this failure here when the parent has an issue?

Reply:
> Why is it not necessary to demonitor before the worker goes away?

Because either worker or parent process will be terminated in a jiffy. In worse case, worker will just have one `'DOWN'` unprocessed message in its mailbox. We can `demonitor` parent pid with `[flush]` in `after` section, but basically it doesn't change anything.

> Should we really emit this failure here when the parent has an issue?

Hmm, it's a good question. It's not an upstream dns server issue, so probably no.